### PR TITLE
Add the ability to pass a value with the -d flag

### PR DESCRIPTION
**Pull request by Jeff-Ciesielski**
This allows the end user to use the {.intdefine/strdefine.}
pragmas to pass values into code at compile time.  This has a nice side
effect of also allowing/requiring a default value to be assigned in the
code (see osalloc.nim/StandaloneHeapSize for an example)

### CODE REVIEW

**Comment by vegansk** [comment]
+100500 votes for this PR!

**Comment by Varriount** [comment]
@Jeff-Ciesielski Overall I think this would be a welcome feature, however I have a few remarks.

First, why not only accept string values? Couldn't those always be parsed with procedures from strutils? On a similar note, why no floatdefine?

Second, an addition to the manual and news files is needed for such a new feature.

File: compiler/pragmas.nim [hunk 1]
```diff
@@ -898,6 +899,10 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: int,
       of wBase:
         noVal(it)
         sym.flags.incl sfBase
+      of wIntDefine:
+        sym.magic = mIntDefine
+      of wStrDefine:
+        sym.magic = mIntDefine
```
Comment (kierdavis) [inline compiler/pragmas.nim]: Should this say `sym.magic = mStrDefine`?

File: compiler/commands.nim [hunk 2]
```diff
@@ -342,7 +349,11 @@ proc processSwitch(switch, arg: string, pass: TCmdLinePass, info: TLineInfo) =
     discard "allow for backwards compatibility, but don't do anything"
   of "define", "d":
     expectArg(switch, arg, pass, info)
-    defineSymbol(arg)
+    if hasKeyValuePair(arg):
```
Comment (Araq) [inline compiler/commands.nim]: ``if {':','='} in arg``

File: compiler/commands.nim [hunk 3]
```diff
@@ -122,6 +122,13 @@ proc splitSwitch(switch: string, cmd, arg: var string, pass: TCmdLinePass,
   elif switch[i] in {':', '=', '['}: arg = substr(switch, i + 1)
   else: invalidCmdLineOption(pass, switch, info)
 
+proc hasKeyValuePair(arg: string): bool =
```
Comment (Araq) [inline compiler/commands.nim]: Completely unnecessary proc.

File: compiler/pragmas.nim (hunk 1 above)
Comment (Araq) [inline compiler/pragmas.nim]: Indeed, typo here.

**Comment by Araq** [comment]
This solution is excellent.

**Comment by Jeff-Ciesielski** [comment]
@Varriount: It's been my experience that passing integer values in at build time is the more frequent usecase (max rtos threads / max heap size / etc), though I have made use of strings for version printouts and whatnot.  Seemed a small amount of overhead to provide both and will make sure their use is consistent.  As for your mention of floatdefine, If you feel strongly about it, I'd be happy to toss it in there before inclusion.

**Pull request merged**